### PR TITLE
feat: walletconnect wallet

### LINKED ISSUE #1686: WalletConnect integration
### Overview

We want to be able to connect *other wallets* into ShapeShift web app via WalletConnect, i.e. a user has another wallet, and we are the dapp.

*Using the ShapeShift native wallet, as a wallet to connect to other dapps, is out of scope.*

### References and additional details

* Spec here https://www.notion.so/shapeshift/WalletConnect-part-1-Integration-in-v2-e56a2b444132497595c201aef3e18500
* Depends on https://github.com/shapeshift/hdwallet/issues/499 and will block completion
* We can probably implement/comment out/put a half baked implementation behind a flag

### Acceptance Criteria

* Feature flag added
* Can connect via WalletConnect

### Need By Date

_No response_

### Screenshots/Mockups

_No response_

### Estimated effort

1 day

### PR DESCRIPTION
**Pull request by GMSteuart**
## Description

Integrates WalletConnect wallet

UPDATE: Moved the provider object from hdwallet over to web since there is a scenario where the connect modal will hang indefinitely and there is no graceful way to automatically exit on behalf of the user. This indefinite situation occurs after the QR modal pops up and the user rejects the session. The reason this occurs is because the connector at the time is only listening for `connect` events and the only event emitted is the `disconnect` event; alas, the connector enters a state of waiting with no apparent end. The code of concern can be found in WalletConnect's [web3-provider]( https://github.com/WalletConnect/walletconnect-monorepo/blob/7573fa9e1d91588d4af3409159b4fd2f9448a0e2/packages/providers/web3-provider/src/index.ts#L227).

## Notice

- [x] Have you followed the guidelines in our [Contributing]('https://github.com/shapeshift/web/CONTRIBUTING.md) guide?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/shapeshift/web/pulls) for the same update/change?

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

Closes #1686 

## Risk

- UX exposure to WalletConnect bugs

## Testing

Open a terminal for web and another for hdwallet.

```sh
# /hdwallet
yarn
yarn build
cd packages/hdwallet-walletconnect
yarn link
# hdwallet-walletconnect tests
 yarn test packages/hdwallet-walletconnect/src/walletconnect.test.ts
#  hdwallet-walletconnect integration tests
cd integration
yarn test src/walletconnect.test.ts
```

```sh
# /web
# clone this branch 
gh pr checkout 1733
yarn clean
yarn link "@shapeshiftoss/hdwallet-walletconnect"
# update package.json "@shapeshiftoss/hdwallet-walletconnect": "^1.21.2"
# to use filepath "@shapeshiftoss/hdwallet-walletconnect": "../hdwallet/packages/hdwallet-walletconnect"
# the above assumes your directory structure is setup like:
# .
# ├── hdwallet
# ├── lib
# └── web
yarn
open https://test.walletconnect.org/
yarn dev
```

1. Sign in with some other wallet
2. Go to flag settings
3. Turn on WalletConnectWallet flag and save
4. Disconnect from current wallet
5. Go through wallet sign in process but select WalletConnect this time around

## Screenshots (if applicable)

Screen Shot 2022-05-18 at 3 33 42 PM

Screen Shot 2022-05-18 at 3 34 36 PM

Screen Shot 2022-05-18 at 3 35 26 PM

@0xdef1cafe copy on pairing modal will need to be updated since a QR code pops up:

Screen Shot 2022-05-18 at 3 35 59 PM

Screen Shot 2022-05-12 at 3 41 20 PM

Screen Shot 2022-05-18 at 3 43 58 PM

On [test.walletconnect.org](test.walletconnect.org)
Screen Shot 2022-05-18 at 3 48 27 PM

Screen Shot 2022-05-18 at 3 51 06 PM

And lastly on [test.walletconnect.org](test.walletconnect.org)
Screen Shot 2022-05-18 at 3 51 53 PM

### CODE REVIEW

**Comment by GMSteuart** [comment]
@0xdef1cafe will need product review for user flow when clicking WalletConnect option

**Comment by gomesalexandre** [comment]
Also, seems like the diff touches many files/lines that shouldn't be modified by this PR, rebase/merge might help 👀

**Comment by GMSteuart** [comment]
> @GMSteuart I suppose this needs linking with [shapeshift/hdwallet#529](https://github.com/shapeshift/hdwallet/pull/529) for now, could you add some instructions for local testing until `hdwallet-walletconnect` is published?

Ya, I can add local testing steps, one of those will have to be running the associated hdwallet-walletconnect package that isn't published yet.

**Comment by GMSteuart** [comment]
> Also, seems like the diff touches many files/lines that shouldn't be modified by this PR, rebase/merge might help 👀

Oh fun. lol. Glad I'm finding out earlier in the day rather than later. Thanks for pointing that out. 🙂

File: package.json [hunk 1]
```diff
@@ -24,6 +24,7 @@
     "@shapeshiftoss/hdwallet-native": "^1.21.1",
     "@shapeshiftoss/hdwallet-native-vault": "^1.21.1",
     "@shapeshiftoss/hdwallet-portis": "^1.21.1",
+    "@shapeshiftoss/hdwallet-walletconnect": "/Users/gmsteuart/Workspace/shapeshift/hdwallet/packages/hdwallet-walletconnect",
```
Comment (GMSteuart) [inline package.json]: @gomesalexandre Here's an example.

File: src/Routes/RoutesCommon.tsx [hunk 2]
```diff
@@ -150,7 +150,9 @@ export const routes: Array<NestedRoute> = [
     path: '/flags',
     label: 'navBar.featureFlags',
     icon: <FaFlag />,
-    hide: window.location.hostname !== 'localhost',
+    hide:
+      window.location.hostname !== 'localhost' &&
+      window.location.hostname !== process.env.REACT_APP_LOCAL_IP,
```
Comment (0xdef1cafe) [inline src/Routes/RoutesCommon.tsx:155]: this won't work - `process.env` is not available in the browser

File: src/assets/translations/en/main.json [hunk 3]
```diff
@@ -958,6 +958,21 @@
       "body": "By connecting a wallet, you agree to ShapeShift's Terms of Service and acknowledge that you have read and understand the ShapeShift disclaimer.",
       "footer": "Don't have a wallet?",
       "create": "Create one"
+    },
+    "walletConnect": {
+      "errors": {
```
Comment (gomesalexandre) [inline src/assets/translations/en/main.json:963]: This is missing the `rpcUrlNotFound` translation

File: src/context/WalletProvider/WalletConnect/components/Connect.tsx [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    setError(e)
+    setLoading(false)
+  }
+
+  const pairDevice = async () => {
+    setError(null)
+    setLoading(true)
+
+    const rpcUrl = getConfig().REACT_APP_ETHEREUM_NODE_URL
+    if (!rpcUrl) {
```
Comment (gomesalexandre) [inline src/context/WalletProvider/WalletConnect/components/Connect.tsx:43]: I'm wondering if this case is ever hit: Tried adding a debugger before that block, and if `REACT_APP_ETHEREUM_NODE_URL` isn't present in config, `EnvMissingError: REACT_APP_ETHEREUM_NODE_URL` is thrown by `getConfig()`, and the execution stops there for `pairDevice()`.

This is probably a logic coming over from previous `process.env.X_Y_Z` checks, where indeed, an undefined env var would result in an `undefined` expression, whereas using `getconfig` results in a thrown error.

File: src/context/WalletProvider/WalletConnect/components/Connect.tsx [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+        const deviceId = await wallet.getDeviceID()
+
+        dispatch({
+          type: WalletActions.SET_WALLET,
+          payload: { wallet, name, icon, deviceId },
+        })
+        dispatch({ type: WalletActions.SET_IS_CONNECTED, payload: true })
+        setLocalWalletTypeAndDeviceId(KeyManager.WalletConnect, deviceId)
+        dispatch({ type: WalletActions.SET_WALLET_MODAL, payload: false })
+      } catch (e: any) {
```
Comment (gomesalexandre) [inline src/context/WalletProvider/WalletConnect/components/Connect.tsx:72]: Non-blicking but in general, happier with the error being typed as `unknown` and the right guards/checks being applied to narrow down error type

**Review by gomesalexandre — COMMENTED** [review]
Tested with Trust Wallet and MM mobile, as well as https://test.walletconnect.org/ and LGTM.
Didn't try to broadcast any Tx nor do any advanced regression testing (trading etc) since MM desktop isn't supported by WalletConnect.

Added a few comments WRT raw translation strings, possibly unreachable code.

File: src/assets/translations/en/main.json (hunk 3 above)
Comment (pastaghost) [reply src/assets/translations/en/main.json:963]: 👍

File: src/context/WalletProvider/WalletConnect/components/Connect.tsx (hunk 4 above)
Comment (pastaghost) [reply src/context/WalletProvider/WalletConnect/components/Connect.tsx:43]: Looking at this now, I agree - this code isn't reachable. There is no such check in `metamask/Connect.tsx` either, so I'll remove these lines here.

File: src/context/WalletProvider/WalletConnect/components/Connect.tsx (hunk 5 above)
Comment (pastaghost) [reply src/context/WalletProvider/WalletConnect/components/Connect.tsx:72]: 👍

**Review by gomesalexandre — DISMISSED** [review]
Smallish comments but otherwise looks good, tested again locally and I was able to connect with WalletConnect.
Did a general regression testing and I didn't find any regression nor weird console logs/errors!

🍬

File: src/assets/translations/en/main.json [hunk 6]
```diff
@@ -843,6 +843,22 @@
         "body": "Unable to connect Portis wallet"
       }
     },
+    "walletConnect": {
+      "errors": {
+        "unknown": "An unexpected error occurred communicating with WalletConnect",
+        "connectFailure": "Unable to connect WalletConnect wallet",
+        "rpcUrlNotFound": "Please configure REACT_APP_ETHEREUM_NODE_URL in your .env file to point to your Ethereum node"
```
[the remark this answers is not in the document]
Comment (pastaghost) [reply src/assets/translations/en/main.json:850]: 👍

File: src/context/WalletProvider/WalletConnect/components/Connect.tsx [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+  extends RouteComponentProps<
+    {},
+    any, // history
+    LocationState
+  > {
+  dispatch: React.Dispatch<ActionTypes>
+}
+
+const moduleLogger = logger.child({
+  namespace: ['WalletConnect'],
```
[the remark this answers is not in the document]
Comment (pastaghost) [reply src/context/WalletProvider/WalletConnect/components/Connect.tsx:26]: 👍

File: src/context/WalletProvider/WalletConnect/components/Connect.tsx [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+        })
+        dispatch({ type: WalletActions.SET_IS_CONNECTED, payload: true })
+        setLocalWalletTypeAndDeviceId(KeyManager.WalletConnect, deviceId)
+        dispatch({ type: WalletActions.SET_WALLET_MODAL, payload: false })
+      } catch (e: unknown) {
+        if (e instanceof Error) {
+          if (e?.message?.startsWith('walletProvider.')) {
+            moduleLogger.error(
+              { e },
+              'WalletConnect Connect: There was an error initializing the wallet',
```
[the remark this answers is not in the document]
Comment (pastaghost) [reply src/context/WalletProvider/WalletConnect/components/Connect.tsx:81]: 👍

File: src/context/WalletProvider/WalletConnect/components/Connect.tsx [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+  // The WalletConnect modal handles desktop and mobile detection as well as deep linking
+  return (
+    <ConnectModal
+      headerText={'walletProvider.walletConnect.connect.header'}
+      bodyText={'walletProvider.walletConnect.connect.body'}
+      buttonText={'walletProvider.walletConnect.connect.button'}
+      pairDevice={pairDevice}
+      loading={loading}
+      error={error}
+    ></ConnectModal>
```
[the remark this answers is not in the document]
Comment (pastaghost) [reply src/context/WalletProvider/WalletConnect/components/Connect.tsx:103]: 👍

File: src/context/WalletProvider/WalletConnect/components/Failure.tsx [hunk 10]
```diff
@@ -0,0 +1,10 @@
+import { FailureModal } from 'context/WalletProvider/components/FailureModal'
+
+export const WalletConnectFailure = () => {
+  return (
+    <FailureModal
+      headerText={'walletProvider.walletConnect.failure.header'}
+      bodyText={'walletProvider.walletConnect.failure.body'}
+    ></FailureModal>
```
[the remark this answers is not in the document]
Comment (pastaghost) [reply src/context/WalletProvider/WalletConnect/components/Failure.tsx:8]: 👍

File: package.json [hunk 11]
```diff
@@ -18,6 +18,7 @@
     "@shapeshiftoss/caip": "^5.0.1",
     "@shapeshiftoss/chain-adapters": "^4.6.0",
     "@shapeshiftoss/errors": "^1.1.1",
+    "@shapeshiftoss/hdwallet-walletconnect": "^1.23.1-alpha.0",
```
[the remark this answers is not in the document]
Comment (pastaghost) [reply package.json]: 👍

**Review by gomesalexandre — APPROVED** [review]

File: README.md [hunk 12]
```diff
@@ -10,12 +10,20 @@ ShapeShift's OSS 2nd generation Web application. (Under Development)
 
 ## Table Of Contents
 
-- [Helpful Docs](#helpful-docs)
-- [Resources](#resources)
-- [Dependencies](#dependencies)
-- [Developer Onboarding](#developer-onboarding)
-- [Commands](#commands)
-- [Linking local dependencies](#linking)
+- [ShapeShift DAO Web Interface](#shapeshift-dao-web-interface)
```
Comment (0xdef1cafe) [inline README.md:13]: separate PR please

File: package.json [hunk 13]
```diff
@@ -98,6 +98,7 @@
     "@shapeshiftoss/hdwallet-native-vault": "^1.23.0",
     "@shapeshiftoss/hdwallet-portis": "^1.23.0",
     "@shapeshiftoss/hdwallet-tallyho": "^1.22.0",
+    "@shapeshiftoss/hdwallet-walletconnect": "^1.23.1-alpha.0",
```
Comment (0xdef1cafe) [inline package.json:101]: requesting changes so we don't merge this as is

**Review by 0xdef1cafe — CHANGES_REQUESTED** [review]
can't publish this with an alpha package

**Comment by 0xdef1cafe** [comment]
image

i connected a wallet with eth, btc, and atom assets

i can see tx history for eth and atom but not bitcoin

i can only see account data for eth

i'm only seeing eth account specifiers in the client

image

plz fix

**Comment by GMSteuart** [comment]
Screen Shot 2022-06-09 at 6 12 49 PM

@0xdef1cafe no longer a fuzzy fox 🙂

**Comment by pastaghost** [comment]
> i connected a wallet with eth, btc, and atom assets
i can see tx history for eth and atom but not bitcoin
i can only see account data for eth
i'm only seeing eth account specifiers in the client

I think this is an hdwallet issue. I'll make the necessary changes and publish a new alpha release, then include the updated dependency here.

**Comment by gomesalexandre** [comment]
This is currently blocked by the hdwallet WalletConnect integration: https://github.com/shapeshift/hdwallet/pull/544

File: src/components/Modals/Send/views/Details.tsx [hunk 14]
```diff
@@ -136,7 +141,11 @@ export const Details = () => {
                   {cryptoSymbol}
                 </Button>
               }
-              inputRightElement={<SendMaxButton onClick={handleSendMax} />}
+              inputRightElement={
+                !(wallet?.getVendor() === 'WalletConnect') ? (
```
Comment (gomesalexandre) [inline src/components/Modals/Send/views/Details.tsx:145]: Prefer `x !== y` vs. `!(x === z)`, no reason not to go for the arithmetical not equal unless you're doing multiple not equal checks and want to use only one `!` for cleanness

**Review by gomesalexandre — CHANGES_REQUESTED** [review]
## Trust Wallet (derived ETH Account with ETH only):

- Accounts/Account/Asset pages are properly working and showing the wallet ETH amount
- Portfolio balance is properly shown
- Tried an ETH self-send which properly went through, however, I'm getting a "User rejected the transaction" toast
- Can't see any Tx history

## MetaMask Wallet (derived ETH Account with ETH and ERC-20s):

- Accounts/Account/Asset pages are properly working and showing the wallet ETH and ERC-20s amount
- Portfolio balance is properly shown
- Tried an ETH self-send which properly went through, however, I'm getting a "User rejected the transaction" toast
- Can't see any Tx history

## Keplr Mobile

Keplr also supports WalletConnect, and can be used to connect to e.g osmosis.zone. It isn't supported at all as the app is effectively in full skeleton mode for dashboard and has no loaded accounts in state.
Do we want to handle the case of Keplr Mobile (and Cosmos/Osmosis accounts derivation in WalletConnect) not being supported, by having some UX for it?

File: src/components/Modals/Send/views/Details.tsx [hunk 15]
```diff
@@ -136,7 +141,11 @@ export const Details = () => {
...
+                  <SendMaxButton onClick={handleSendMax} />
+                ) : undefined
```
[the remark this answers is not in the document]
Comment (pastaghost) [reply src/components/Modals/Send/views/Details.tsx:147]: 👍

File: src/components/Trade/TradeInput.tsx [hunk 16]
```diff
@@ -289,17 +289,19 @@ export const TradeInput = ({ history }: RouterProps) => {
...
+                      size='sm'
+                      variant='ghost'
+                      colorScheme='blue'
+                      isDisabled={isSendMaxLoading || !hasValidBalance || !quote}
+                      onClick={onSetMaxTrade}
+                      data-test='token-row-sell-max-button'
+                    >
+                      Max
+                    </Button>
+                  ) : undefined
```
[the remark this answers is not in the document]
Comment (pastaghost) [reply src/components/Trade/TradeInput.tsx:304]: 👍

**Comment by pastaghost** [comment]
> Keplr also supports WalletConnect, and can be used to connect to e.g osmosis.zone. It isn't supported at all as the app is effectively in full skeleton mode for dashboard and has no loaded accounts in state.
Do we want to handle the case of Keplr Mobile (and Cosmos/Osmosis accounts derivation in WalletConnect) not being supported, by having some UX for it?

This is because WalletConnect 1.0 (the version we've implemented in HDWallet) is ETH-only. WalletConnect 2.0 has multi-chain support but is still in beta. I'm currently looking to see if it is possible to either detect the user wallet connected on the other side of the WalletConnect bridge or to limit the set of wallets supported by the application via WalletConnect.

File: src/components/Modals/Send/views/Details.tsx [hunk 17]
```diff
@@ -142,9 +142,9 @@ export const Details = () => {
                 </Button>
               }
               inputRightElement={
-                !(wallet?.getVendor() === 'WalletConnect') ? (
+                wallet?.getVendor() !== 'WalletConnect' ? null : (
```
Comment (gomesalexandre) [inline src/components/Modals/Send/views/Details.tsx:145]: `===`?

File: src/components/Modals/Send/views/Details.tsx [hunk 18]
```diff
@@ -171,9 +171,9 @@ export const Details = () => {
...
```
Comment (gomesalexandre) [inline src/components/Modals/Send/views/Details.tsx:174]: `===`?

**Review by gomesalexandre — DISMISSED** [review]
Tested it again and managed to get connect, disconnect, and Txs (swaps/sends) working. Note that I didn't continue with actually broadcasting Txs and rejected the Tx confirm popup.

Only thing I've noticed is a little styling issue, we're missing right padding on the WalletConnect option but we can tackle it as a smallish follow-up. LGTM otherwise!

image

Approving this one even though this is still waiting for a product copy update, I'll re-stamp after. Logic looks good with the latest commits, good job ser!

🍬

**Comment by pastaghost** [comment]
* Added copy to WalletConnect option in `SelectModal.tsx`
* Constrained width of wallet option icons
* Tested WalletConnect option text in all three possible configurations (no wallet active, walletConnect active, some other wallet active)

Screen Shot 2022-07-19 at 3 59 33 PM
Screen Shot 2022-07-19 at 3 59 16 PM
Screen Shot 2022-07-19 at 3 58 52 PM

**Review by gomesalexandre — DISMISSED** [review]
Tested again with the latest changes, both the "Ethereum only" copy and WalletConnect option button look good.

🍬

**Review by 0xdef1cafe — APPROVED** [review]

**Comment by 0xdef1cafe** [comment]
merging as previously approved by @gomesalexandre and the most recent commit was stylistic/trivial

**Pull request merged**